src/main.c

```diff
@@ -256,9 +304,14 @@ poe_cmd_queue(unsigned char *_cmd, int len)
 
 	list_add_tail(&cmd->list, &cmd_pending);
 
-	if (empty)
+	if (!empty)
+		return 0;
+	if (config.i2c_bus < 0)
```

Comment:
The way that this case statement has been rewritten makes it very difficult to suss out what's special about i2c vs. uart reply handling.

After a few re-readings, I understand that

- this does not change control flow for uart, and
- for i2c, as long as there are no items in `cmd_pending` at the start of `poe_cmd_queue`, then we continue and invoke `poe_i2c_cmd_send`.

`poe_cmd_queue` is the only place where `poe_i2c_cmd_send`. This means that the i2c implementation has hijacked the 'queue': it will only carry the command being sent at any point in time. Should `cmd_pending` ever contain anything else, this will never complete, as `poe_i2c_cmd_send` can never process.

You haven't just implemented i2c more synchronously, you've also watered down the functionality of the command queue. Even if this patch works, it means we cannot rely on `cmd_pending` to work as a queue later down the line (e.g. if we want to enqueue a packet but not process it right now, it'll break i2c).

Please either re-implement i2c through ustream, or design / wait for an implementation of reply queueing that doesn't use it. I understand that it will be less efficient (because `ustream` `read()`s instead of using the `I2C_RDWR` ioctl). I'll be trying to work out the latter option, but my understanding of C threading is abysmal.

Reply:
Not ustream. jow points out that that's a horrible idea; use e.g.

```txt
 maybe do the required EAGAIN / EINTR nonblock IO repeat bilerplate around
 struct my_stuff { struct uloop_fd ufd; other stuff };
 struct my_stuff *blah = calloc(1, sizeof(struct my_stuff));
 my_stuff->ufd.fd = socket(...); // or open() or whatever you gyus do there
 my_stuff->...; // init other stuff you might need
 my_stuff->ufd.cb = my_read_callback;
 uloop_fd_add(&my_stuff->ufd, ULOOP_READ);
 in your callback:
 struct my_stuff *blah = container_off(ufd /* the uloop_fd pointer you get as 1st arg in the callback */, struct my_stuff, ufd);
 *container_of() 
 ```

Reply:
I would prefer to have a threaded implementation that entirely hides I2C and UART behind it. They would both be synchronous, and the thread would provide the asynchronicity. There would be a thread for UART and one for I2C depending on what is needed. We could also handle command repeats in the thread so that the queueing is cleaner. The queue would be the interface between the asynchronous front-end and a synchronous backend which also handles some of the error cases, like command repeats. After all it would know better about timing.
I have done some heavy pthread stuff in the past, but I have not touched it since more than a decade, and I would probably need to start again at more-or-less zero.

But first I need to get the PoE going at all, the failing commands lead to the power not coming up properly.

Reply:
For the failing commands, please try to rebase on #15 .

I have some ideas where to snip the interface. As long as the backend gets its commands from the `cmd_pending` queue, and puts the replies in via `poe_reply_consume()`, that's the basis of the interface. Whether the backend runs in a thread, or as part of uloop() with epoll(), it doesn't matter. So there would be no need to do a thread for UART.

I'll try to get a basic split of UART code in the next few days. There's some context tracking and other boilerplate that I might need to implement.

Reply:
The error handling comes too late for the I2C code and it also does not apply to the checksum errors. If there is a checksum error, the code never gets into `handle_poe_f0_reply()` in `poe_reply_consume()`.
What works for me is the following:
```
static int
poe_cmd_queue(unsigned char *_cmd, int len)
{
	int i, empty = list_empty(&cmd_pending);
	struct cmd *cmd = malloc(sizeof(*cmd));
	unsigned char reply[12];
	unsigned char sum;
	int retries = 10;

	memset(cmd, 0, sizeof(*cmd));
	memset(cmd->cmd, 0xff, 12);
	memcpy(cmd->cmd, _cmd, len);

	cmd_seq++;
	cmd->cmd[1] = cmd_seq;
	cmd->cmd[11] = 0;

	for (i = 0; i < 11; i++)
		cmd->cmd[11] += cmd->cmd[i];

	list_add_tail(&cmd->list, &cmd_pending);

	if (!empty)
		return 0;
	if (config.i2c_bus < 0)
		return poe_cmd_send(cmd);

	do {
		poe_i2c_cmd_send(cmd, reply, sizeof(reply));
		sum = 0;
		for (i = 0; i < 11; i++)
			sum += reply[i];
		if (reply[11] != sum) {
			ULOG_DBG("received reply with bad checksum, RETRY\n");
		} else {
			break;
		}
		usleep(10000);
	} while (--retries);

 	poe_reply_consume(reply);

	return 0;
}
```
That's why I would suggest to handle checksum errors with a simple retry in a thread that handles the top of the command queue.

Using the above code the all FF responses are no longer an issue. It seems indeed as if the MCU is not ready and repeating the command after 10ms seems always to succeed.

Reply:
Ohh, the all F's reply has an invalid checksum on I2C. Fun!

Feel free to develop the I2C path so that you can power on the ports. The goal is to obtain the knowledge on how to drive the MCU to make ports work. The final code will likely change many times, as we split the backends.